_posts/2022-06-15-inline-assembly-memory-side-effects-bug.md

```diff
@@ -0,0 +1,156 @@
+under certain conditions.
+
+The bug was introduced in Solidity 0.8.13 with a new Yul optimizer step meant to
+remove unused stores to memory and storage.
+
+We assigned the bug a severity of "medium".
+
+## Which Contracts are Affected?
+
+The Yul optimizer considers all memory writes in the outermost Yul block that are
+never read from as unused and removes them. This is valid when that Yul block is
+the entire Yul program, which is always the case for the Yul code generated by the
+new via-IR pipeline. Inline assembly blocks are never optimized in isolation when
+using that pipeline. Instead they are optimized as a part of the whole Yul input.
+
+However, the legacy code generation pipeline (which is still the default) runs the
+Yul optimizer individually on an inline assembly block if the block does not refer
+to any local variables defined in the surrounding Solidity code. Consequently, memory
+writes in such inline assembly blocks are removed as well, if the written memory is
+never read from in the same assembly block, even if the written memory is accessed
+later, for example by a subsequent inline assembly block.
+
+Fortunately, the fact that the legacy code generation pipeline does not run the Yul
+optimizer at all on inline assembly blocks that do access local Solidity variables,
+reduces the number of affected cases significantly. Most inline assembly blocks either
+read or write values from or to local variables of the surrounding Solidity code, or
+are entirely self-contained, or take over the program flow up until to the end of
+the transaction. Thereby, the impact of the bug is likely low in practice. However,
+since the behaviour in affected cases is thoroughly unexpected, we assigned a severity
+of "medium".
+
+
+In the following example, the legacy code generation pipeline with enabled optimizer will
+remove the ``mstore`` and the function ``f`` will return zero:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+        }
+        assembly {
+            x := mload(0)
+        }
+    }
+}
+```
+
+However, if the same memory is either read again in the same inline assembly block
+or if the inline assembly block accesses any local Solidity variables, the bug is
+not present. Both is the case in the following example, so ``f`` will return
+``0x42`` as expected:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+            x := mload(0)
+        }
+    }
+}
+```
+
+In the following example, the first ``mstore`` will not be removed, since the written memory
+is read again by the ``return``. The second ``mstore`` on the other hand will be removed, since the memory
+it writes is never read again. In this case, this is a valid optimization and the example is not
+affected by the bug. Generally, any assembly block that terminates or does not have memory side-effects that
```

Comment:
```suggestion
adversely affected by the bug. Generally, any assembly block that terminates the execution of the whole contract or does not have memory side-effects that
```

Reply:
I'll say "the transaction".

Reply:
Sounds good too. I originally wanted to say "the program" but contract just sounded better to me.